File: ros2_interfaces/src/geomagic_control_py/geomagic_control_py/full_haptic.py
import rclpy
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node

from std_srvs.srv import SetBool
from sensor_msgs.msg import JointState
from geometry_msgs.msg import Wrench, Pose
from std_msgs.msg import Int32MultiArray
from rcl_interfaces.msg import ParameterDescriptor, ParameterType

from ABBRobotEGM import EGM
import PyKDL


NODE = 'haptic_cartesian_bridge'
HAPTIC_NODE_NAME = 'haptic_device_ros2'


# ===================== AUX FUNCTIONS =====================
def from_msg_pose(msg_pose, kdl_frame):
    kdl_frame.p = PyKDL.Vector(
        msg_pose.position.x,
        msg_pose.position.y,
        msg_pose.position.z
    )
    kdl_frame.M = PyKDL.Rotation.Quaternion(
        msg_pose.orientation.x,
        msg_pose.orientation.y,
        msg_pose.orientation.z,
        msg_pose.orientation.w
    )


def from_state_abb(state, kdl_frame):
    kdl_frame.p = PyKDL.Vector(
        state.cartesian.pos.x * 0.001,
        state.cartesian.pos.y * 0.001,
        state.cartesian.pos.z * 0.001
    )
    kdl_frame.M = PyKDL.Rotation.Quaternion(
        state.cartesian.orient.u1,
        state.cartesian.orient.u2,
        state.cartesian.orient.u3,
        state.cartesian.orient.u0
    )


def to_state_abb(kdl_frame):
    pos_mm = [
        kdl_frame.p.x() * 1000.0,
        kdl_frame.p.y() * 1000.0,
        kdl_frame.p.z() * 1000.0
    ]
    qx, qy, qz, qw = kdl_frame.M.GetQuaternion()
    orient = [qw, qx, qy, qz]

    return pos_mm, orient


# ===================== NODE =====================
class HapticCartesianBridge(Node):

    def __init__(self):
        super().__init__(NODE)

        # -------- STATE --------
        self.first_haptic_output = False
        self.first_robot_output = False
        self.capture_enabled = True

        self.H_0_N_sensor_initial = PyKDL.Frame()
        self.H_0_N_robot_initial = PyKDL.Frame()
        self.H_N_robot_0_sensor = PyKDL.Frame()

        self.scale_ = 1.0

        # -------- INIT --------
        self.set_parameters()
        self.create_egm()
        self.create_haptic_interface()
        self.create_subscribers()

        self.get_logger().info('Unified Haptic-Cartesian Bridge started')

    # ===================== HAPTIC =====================
    def create_haptic_interface(self):
        # Publisher feedback
        self.feedback_pub = self.create_publisher(
            JointState,
            HAPTIC_NODE_NAME + '/feedback',
            10
        )

        # Set feedback mode
        self.client = self.create_client(
            SetBool,
            HAPTIC_NODE_NAME + '/set_feedback_mode'
        )

        req = SetBool.Request()
        req.data = False

        while not self.client.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('Waiting for haptic service...')

        future = self.client.call_async(req)
        rclpy.spin_until_future_complete(self, future)

        if future.result() is None:
            raise RuntimeError("Failed to set feedback mode")

    # ===================== EGM =====================
    def create_egm(self):
        self.abb = EGM()

        for _ in range(10):
            success, state = self.abb.receive_from_robot()
            if success:
                self.cartesian_initial_pose(state)
                break
        else:
            raise RuntimeError("No EGM connection")

        self.get_logger().info("EGM initialized")

    # ===================== PARAMETERS =====================
    def set_parameters(self):
        self.declare_parameter('scale', 1.0)
        self.scale_ = self.get_parameter('scale').value

        self.declare_parameter('roll_N_sensor', 0.0)
        self.declare_parameter('pitch_N_sensor', 0.0)
        self.declare_parameter('yaw_N_sensor', 0.0)

        r = self.get_parameter('roll_N_sensor').value
        p = self.get_parameter('pitch_N_sensor').value
        y = self.get_parameter('yaw_N_sensor').value

        self.H_N_robot_0_sensor.M = PyKDL.Rotation.RPY(r, p, y)

    # ===================== SUBSCRIBERS =====================
    def create_subscribers(self):
        self.create_subscription(Wrench, '/state', self.force_callback, 10)
        self.create_subscription(Pose, HAPTIC_NODE_NAME + '/state/pose', self.pose_callback, 10)
        self.create_subscription(Int32MultiArray, HAPTIC_NODE_NAME + '/state/buttons', self.buttons_callback, 10)

    # ===================== CALLBACKS =====================
    def force_callback(self, msg):
        self.feedback_pub.publish(msg)

    def buttons_callback(self, msg):
        if len(msg.data) < 2:
            return

        if msg.data[0] == 1:
            self.first_robot_output = False

        if msg.data[1] == 1:
            self.capture_enabled = not self.capture_enabled

    def pose_callback(self, msg):
        if not self.capture_enabled:
            return

        self.set_initial_haptic(msg)

        if not (self.first_haptic_output and self.first_robot_output):
            return

        cmd = self.compute_motion(msg)
        self.send_to_robot(cmd)

    # ===================== LOGIC =====================
    def set_initial_haptic(self, msg):
        if not self.first_haptic_output:
            from_msg_pose(msg, self.H_0_N_sensor_initial)
            self.first_haptic_output = True

    def cartesian_initial_pose(self, state):
        if not self.first_robot_output:
            from_state_abb(state, self.H_0_N_robot_initial)
            self.first_robot_output = True

    def compute_motion(self, msg):
        current = PyKDL.Frame()
        from_msg_pose(msg, current)

        delta = self.H_0_N_sensor_initial.Inverse() * current
        delta.M = PyKDL.Rotation.Identity()

        p = self.H_0_N_robot_initial * (
            self.H_N_robot_0_sensor * delta.p
        )

        p = PyKDL.Vector(
            p.x() * self.scale_,
            p.y() * self.scale_,
            p.z() * self.scale_
        )

        result = PyKDL.Frame(self.H_0_N_robot_initial)
        result.p = p

        return result

    def send_to_robot(self, frame):
        pos, orient = to_state_abb(frame)
        self.abb.send_to_robot(cartesian=(pos, orient), digital_signal_to_robot=False)


# ===================== MAIN =====================
def main(args=None):
    rclpy.init(args=args)
    node = HapticCartesianBridge()

    try:
        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: ros2_interfaces/src/geomagic_control/src/fullHaptic/fullHaptic.cpp
#include "geomagic_control/fullHaptic/fullHaptic.h"

namespace {
    inline void fromMsg(const geometry_msgs::msg::Pose& in, KDL::Frame& out) {
        out.p = KDL::Vector(in.position.x, in.position.y, in.position.z);
        out.M = KDL::Rotation::Quaternion(in.orientation.x, in.orientation.y,
                                          in.orientation.z, in.orientation.w);
    }

    inline void fromMsg(const geometry_msgs::msg::PoseStamped& in, KDL::Frame& out) {
        out.p = KDL::Vector(in.pose.position.x, in.pose.position.y, in.pose.position.z);
        out.M = KDL::Rotation::Quaternion(in.pose.orientation.x, in.pose.orientation.y,
                                          in.pose.orientation.z, in.pose.orientation.w);
    }

    inline void toMsg(const KDL::Frame& in, geometry_msgs::msg::Pose& out) {
        out.position.x = in.p.x();
        out.position.y = in.p.y();
        out.position.z = in.p.z();
        in.M.GetQuaternion(out.orientation.x, out.orientation.y,
                           out.orientation.z, out.orientation.w);
    }

    inline void toMsg(const KDL::Frame& in, sensor_msgs::msg::JointState& out) {
        out.effort.resize(3);
        out.effort[0] = in.p.x();
        out.effort[1] = in.p.y();
        out.effort[2] = in.p.z();
    }
}

FullHapticNode::FullHapticNode() : Node(NODE_NAME)
{
    publishCartesianCreate();

    if (!setPoseMode()) {
        RCLCPP_ERROR(this->get_logger(), "Fallo en la configuracion pose");
        return;
    }

    if (!setFeedbackMode()) {
        RCLCPP_ERROR(this->get_logger(), "Failed to configure cartesian control server");
        return;
    }

    setParameters();
    hapticSubscribeCreate();

    RCLCPP_INFO(this->get_logger(), "Full Haptic Node started");
    RCLCPP_INFO(this->get_logger(), "Listening to: /haptic_device/state/pose");
    RCLCPP_INFO(this->get_logger(), "Publishing to: /cartesian_control_server_ros2/command/pose");
}

// ------------------------------ Publishers ------------------------------

void FullHapticNode::publishCartesianCreate()
{
    m_cartesian_cmd_pub = this->create_publisher<geometry_msgs::msg::Pose>(
        std::string(CARTESIAN_CONTROL_SERVER_NODE_NAME) + "/command/pose", 10);

    m_haptic_feedback_pub = this->create_publisher<sensor_msgs::msg::JointState>(
        std::string(HAPTIC_YARP_DEVICE_NODE_NAME) + "/feedback", 10);

    client_param_ = create_client<SetParameters>(
        std::string(CARTESIAN_CONTROL_SERVER_NODE_NAME) + "/set_parameters");
}

// ------------------------------ Service clients ------------------------------

bool FullHapticNode::setPoseMode()
{
    using ServiceResponseFuture = rclcpp::Client<SetParameters>::SharedFuture;

    while (!client_param_->wait_for_service(std::chrono::seconds(1))) {
        if (!rclcpp::ok()) {
            RCLCPP_ERROR(get_logger(), "Interrupted while waiting for the service. Exiting.");
            return false;
        }
        RCLCPP_INFO(get_logger(), "Service not available, waiting again...");
    }

    auto request = std::make_shared<SetParameters::Request>();

    Parameter param;
    param.name = "preset_streaming_cmd";
    param.value.type = rclcpp::ParameterType::PARAMETER_STRING;
    param.value.string_value = "pose";
    request->parameters.push_back(param);

    auto response_received_callback = [this](ServiceResponseFuture future) {
        if (future.get()->results[0].successful) {
            RCLCPP_INFO(get_logger(), "Preset streaming command correctly established in external node.");
        } else {
            RCLCPP_ERROR(get_logger(), "Failed to set preset streaming command.");
        }
    };

    auto result = client_param_->async_send_request(request, response_received_callback);
    return result.valid();
}

bool FullHapticNode::setFeedbackMode()
{
    client_set_feedback_mode_ = this->create_client<std_srvs::srv::SetBool>("set_feedback_mode");

    auto request = std::make_shared<std_srvs::srv::SetBool::Request>();
    request->data = false;  // joint torque feedback mode

    while (!client_set_feedback_mode_->wait_for_service(std::chrono::seconds(1))) {
        if (!rclcpp::ok()) {
            RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Interrupted while waiting for the service. Exiting.");
            return false;
        }
        RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "service not available, waiting again...");
    }

    auto result = client_set_feedback_mode_->async_send_request(request);
    if (rclcpp::spin_until_future_complete(this->shared_from_this(), result) ==
        rclcpp::FutureReturnCode::SUCCESS)
    {
        RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "Feedback mode set successfully");
    } else {
        RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Failed to call service set_feedback_mode");
    }

    return result.get()->success;
}

// ------------------------------ Parameters ------------------------------

void FullHapticNode::setParameters()
{
    rcl_interfaces::msg::ParameterDescriptor descriptor_msg;
    // scale
    scale_ = 1.0;
    descriptor_msg.name = "scale";
    descriptor_msg.description = "Scaling factor to apply to the haptic movements (default: 1.0).";
    descriptor_msg.read_only = true;
    descriptor_msg.set__type(rcl_interfaces::msg::ParameterType::PARAMETER_DOUBLE);
    declare_parameter<double>("scale", 1.0, descriptor_msg);
    get_parameter("scale", scale_);

    // roll
    double roll_N_sensor_ = 0.0;
    descriptor_msg.name = "roll_N_sensor";
    descriptor_msg.description = "Rotation about global axis X (roll) from TCP to sensor (radians).";
    descriptor_msg.read_only = true;
    descriptor_msg.set__type(rcl_interfaces::msg::ParameterType::PARAMETER_DOUBLE);
    declare_parameter<double>("roll_N_sensor", 0.0, descriptor_msg);
    get_parameter("roll_N_sensor", roll_N_sensor_);

    // pitch
    double pitch_N_sensor_ = 0.0;
    descriptor_msg.name = "pitch_N_sensor";
    descriptor_msg.description = "Rotation about global axis Y (pitch) from TCP to sensor (radians).";
    descriptor_msg.read_only = true;
    descriptor_msg.set__type(rcl_interfaces::msg::ParameterType::PARAMETER_DOUBLE);
    declare_parameter<double>("pitch_N_sensor", 0.0, descriptor_msg);
    get_parameter("pitch_N_sensor", pitch_N_sensor_);

    // yaw
    double yaw_N_sensor_ = 0.0;
    descriptor_msg.name = "yaw_N_sensor";
    descriptor_msg.description = "Rotation about global axis Z (yaw) from TCP to sensor (radians).";
    descriptor_msg.read_only = true;
    descriptor_msg.set__type(rcl_interfaces::msg::ParameterType::PARAMETER_DOUBLE);
    declare_parameter<double>("yaw_N_sensor", 0.0, descriptor_msg);
    get_parameter("yaw_N_sensor", yaw_N_sensor_);

    H_N_robot_0_sensor.M = KDL::Rotation::RPY(roll_N_sensor_, pitch_N_sensor_, yaw_N_sensor_);
}

// ------------------------------ Subscribers ------------------------------

void FullHapticNode::hapticSubscribeCreate()
{
    m_haptic_posePhantom_sub = this->create_subscription<geometry_msgs::msg::Pose>(
        std::string(HAPTIC_YARP_DEVICE_NODE_NAME) + "/state/pose", 10,
        std::bind(&FullHapticNode::hapticPoseCallback, this, std::placeholders::_1));

    m_haptic_poseTeoRightArm_sub = this->create_subscription<geometry_msgs::msg::PoseStamped>(
        std::string(CARTESIAN_CONTROL_SERVER_NODE_NAME) + "/state/pose", 10,
        std::bind(&FullHapticNode::teoPoseCallback, this, std::placeholders::_1));
}

// ------------------------------ Callbacks ------------------------------

void FullHapticNode::hapticPoseCallback(const geometry_msgs::msg::Pose::SharedPtr msg)
{
    hapticInitialPose(msg);

    RCLCPP_INFO(this->get_logger(),
                "Received haptic pose: [%.3f, %.3f, %.3f]",
                msg->position.x, msg->position.y, msg->position.z);

    comprobateIntialPose(msg);
}

void FullHapticNode::teoPoseCallback(const geometry_msgs::msg::PoseStamped::SharedPtr msg)
{
    teoInitialPose(msg);

    RCLCPP_INFO(this->get_logger(),
                "Received teo Right Arm pose: [%.3f, %.3f, %.3f]",
                msg->pose.position.x, msg->pose.position.y, msg->pose.position.z);

    // --- CartesianToHapticFeedback logic ---
    fromMsg(*msg, current_teo_pose);

    KDL::Vector diff     = current_teo_pose.p - wall_center_;
    double      distance = diff.Norm();

    if (distance > wall_radius_) {
        publishHapticFeedback(KDL::Vector(0, 0, 0));
        return;
    }

    double penetration = wall_radius_ - distance;

    if (penetration > 0) {
        KDL::Vector force_direction = diff / distance;
        double force_magnitude = std::min(force_gain_ * penetration, max_force_);
        KDL::Vector force = force_direction * force_magnitude;
        publishHapticFeedback(force);
    }
}

// ------------------------------ HapticToCartesian helpers ------------------------------

void FullHapticNode::comprobateIntialPose(const geometry_msgs::msg::Pose::SharedPtr msg)
{
    if (!(firstHapticOutput && firstTeoOutput)) {
        RCLCPP_INFO(this->get_logger(),
            "Not recibed intial pose of Haptic %d and Teo %d", firstHapticOutput, firstTeoOutput);
        return;
    }

    geometry_msgs::msg::Pose cartesian_cmd = calculateDiferentialPose(msg);

    m_cartesian_cmd_pub->publish(cartesian_cmd);

    RCLCPP_INFO(this->get_logger(),
                "Published cartesian command: [%.3f, %.3f, %.3f]",
                cartesian_cmd.position.x,
                cartesian_cmd.position.y,
                cartesian_cmd.position.z);
}

geometry_msgs::msg::Pose FullHapticNode::calculateDiferentialPose(
    const geometry_msgs::msg::Pose::SharedPtr msg)
{
    KDL::Frame H_0_N_sensor_current;
    fromMsg(*msg, H_0_N_sensor_current);

    KDL::Frame H_sensor_initial_current = H_0_N_sensor_initial.Inverse() * H_0_N_sensor_current;

    // Solo traslacion
    H_sensor_initial_current.M = KDL::Rotation::Identity();

    // Solo rotacion
    KDL::Frame Rot_0_N_sensor_initial = H_0_N_sensor_initial;
    Rot_0_N_sensor_initial.p = KDL::Vector::Zero();

    auto p = H_0_N_robot_initial * (H_N_robot_0_sensor * (Rot_0_N_sensor_initial * H_sensor_initial_current.p));

    p = KDL::Vector(p.x() * scale_, p.y() * scale_, p.z() * scale_);

    KDL::Frame H_0_N_robot = H_0_N_robot_initial;
    H_0_N_robot.p = p;

    geometry_msgs::msg::Pose out;
    toMsg(H_0_N_robot, out);
    return out;
}

void FullHapticNode::hapticInitialPose(const geometry_msgs::msg::Pose::SharedPtr msg)
{
    if (!firstHapticOutput) {
        fromMsg(*msg, H_0_N_sensor_initial);
        firstHapticOutput = true;
        RCLCPP_INFO(get_logger(), "Initial haptic pose correctly set.");
    }
}

void FullHapticNode::teoInitialPose(const geometry_msgs::msg::PoseStamped::SharedPtr msg)
{
    if (!firstTeoOutput) {
        fromMsg(*msg, H_0_N_robot_initial);
        firstTeoOutput = true;
        RCLCPP_INFO(get_logger(), "Initial teo pose correctly set.");
    }
}

// ------------------------------ CartesianToHapticFeedback helpers ------------------------------

void FullHapticNode::publishHapticFeedback(const KDL::Vector& force)
{
    auto feedback_msg = sensor_msgs::msg::JointState();
    toMsg(KDL::Frame(force), feedback_msg);

    m_haptic_feedback_pub->publish(feedback_msg);

    RCLCPP_WARN_THROTTLE(this->get_logger(), *this->get_clock(), 500,
                         "Collision! Force: [%.2f, %.2f, %.2f] N",
                         force.x(), force.y(), force.z());
}

// ------------------------------ main ------------------------------

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<FullHapticNode>());
    rclcpp::shutdown();
    return 0;
}
